[reland][distributed] pass in timeout to TCP store when initializing

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33434 [reland][distributed] pass in timeout to TCP store when initializing**

Reland of https://github.com/pytorch/pytorch/pull/33325, since the
unit test was flaky and failed on land.

To ensure that the test is not flaky, I bumped the timeout so the rendezvous
does not timeout (timing out the rendezvous in 1s led to the flakiness). I also
generalized our mechanism for retrying on errors to include retrying on errors
due to timeout in rendezvous.

Ran the test 500 times and it all passed, built on MacOS and verified that it passes there too.

Differential Revision: [D19935390](https://our.internmc.facebook.com/intern/diff/D19935390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19935390/)!

rohan-varma committed:
Back out "Revert D19871946: [distributed] pass in timeout to TCP store when initializing"

Reland of https://github.com/pytorch/pytorch/pull/33325, since the
unit test was flaky and failed on land.
To ensure that the test is not flaky, I bumped the timeout so the rendezvous
does not timeout (timing out the rendezvous in 1s led to the flakiness). I also
generalized our mechanism for retrying on errors to include retrying on errors
due to timeout in rendezvous.

Differential Revision: [D19935390](https://our.internmc.facebook.com/intern/diff/D19935390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19935390/)!

[ghstack-poisoned]

pritamdamania87 reviewed (approved):
Please fix lint issues before landing.

rohan-varma committed:
Update on "[reland][distributed] pass in timeout to TCP store when initializing"


Reland of https://github.com/pytorch/pytorch/pull/33325, since the
unit test was flaky and failed on land.

To ensure that the test is not flaky, I bumped the timeout so the rendezvous
does not timeout (timing out the rendezvous in 1s led to the flakiness). I also
generalized our mechanism for retrying on errors to include retrying on errors
due to timeout in rendezvous.

Ran the test 500 times and it all passed, but I will also SSH into the macOS CI docker to ensure that it runs properly there (which is where it failed on)

Differential Revision: [D19935390](https://our.internmc.facebook.com/intern/diff/D19935390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19935390/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[reland][distributed] pass in timeout to TCP store when initializing"


Reland of https://github.com/pytorch/pytorch/pull/33325, since the
unit test was flaky and failed on land.

To ensure that the test is not flaky, I bumped the timeout so the rendezvous
does not timeout (timing out the rendezvous in 1s led to the flakiness). I also
generalized our mechanism for retrying on errors to include retrying on errors
due to timeout in rendezvous.

Ran the test 500 times and it all passed, but I will also SSH into the macOS CI docker to ensure that it runs properly there (which is where it failed on)

Differential Revision: [D19935390](https://our.internmc.facebook.com/intern/diff/D19935390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19935390/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "[reland][distributed] pass in timeout to TCP store when initializing"


Reland of https://github.com/pytorch/pytorch/pull/33325, since the
unit test was flaky and failed on land.

To ensure that the test is not flaky, I bumped the timeout so the rendezvous
does not timeout (timing out the rendezvous in 1s led to the flakiness). I also
generalized our mechanism for retrying on errors to include retrying on errors
due to timeout in rendezvous.

Ran the test 500 times and it all passed, but I will also SSH into the macOS CI docker to ensure that it runs properly there (which is where it failed on)

Differential Revision: [D19935390](https://our.internmc.facebook.com/intern/diff/D19935390/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D19935390/)!

[ghstack-poisoned]

cdzhan commented on torch/distributed/distributed_c10d.py:
```diff
@@ -393,7 +387,9 @@ def init_process_group(backend,
+                init_method, rank, world_size, timeout=timeout
+            )
             store, rank, world_size = next(rendezvous_iterator)
             store.set_timeout(timeout)
```
there is no need to set_timeout again.
